Dashboard announcements feature

kaizoku opened the pull request:
This PR adds an Announcement feature to show announcements on the sidebar of the LMS dashboard. It uses TinyMCE to show an editor in the studio.

**JIRA tickets**: Implements OC-3201

**Screenshots**:

![dashboard announcements](https://user-images.githubusercontent.com/7556571/32713023-53721626-c897-11e7-9514-0abbdb8a89ad.png)
![screenshot-2018-4-24 edit announcements studio](https://user-images.githubusercontent.com/270966/39393694-4314a4f2-4a7d-11e8-9ed2-e60bb83f8c27.png)
![screenshot-2018-4-24 edit announcement studio](https://user-images.githubusercontent.com/270966/39393705-52feb8c6-4a7d-11e8-886b-4c0063154aac.png)
![screenshot-2018-4-24 delete announcement studio](https://user-images.githubusercontent.com/270966/39393703-50623fde-4a7d-11e8-889f-c8ff02c92770.png)


**Sandbox URL**:

* LMS: https://pr16496.sandbox.opencraft.hosting/
* Studio: https://studio-pr16496.sandbox.opencraft.hosting/

**Testing setup**:

1. Enable announcements with `settings.FEATURES['ENABLE_ANNOUNCEMENTS'] = True`.
1. Run migrations `./manage.py cms migrate --settings=devstack`

**Testing instructions**:

1. Login to Studio as a staff member.
1. Navigate to the announcements page in studio: https://studio-pr16496.sandbox.opencraft.hosting/maintenance/announcements/
1. Click the "Create New", add content, ensure announcement is enabled
1. Add more than 5 announcements to check pagination.
1. Login to the LMS as a student.
1. View the Course Dashboard and check that the Announcement is properly displayed
1. Edit the Announcement, disable the active field, and ensure that it is removed from the LMS Dashboard.
1. Delete an announcement and check that the delete endpoint works.

**Author notes and concerns**:

1. I wrote tests for the JSX component using [enzyme](https://github.com/airbnb/enzyme) and [jest](https://facebook.github.io/jest/docs/en/tutorial-react.html), similar to the ones in [Paragon](https://github.com/edx/paragon/blob/master/src/Tabs/Tabs.test.jsx), but I was unable to run them without adding a number of dependencies in package.json which seem out of the scope of this pull request. Are there already plans for adding JSX testing into edx-platform? If desired I could open another PR with those dependencies and configuration options.

**Reviewers**
- [ ] (pomegranited)
- [ ] edX reviewer[s] TBD

**Settings**
```yaml
EDXAPP_FEATURES:
  ENABLE_ANNOUNCEMENTS: true
```

edx-webhook commented:
Thanks for the pull request, @kaizoku! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

<a   href="https://openedx-webhooks.herokuapp.com/github/process_pr?repo=edx%2Fedx-platform&number=16496"   rel="nofollow" >Create an OSPR issue for this pull request</a>.

pomegranited commented on openedx/core/djangoapps/announcements/templates/announcements/announcements_edit.html:
```diff
@@ -0,0 +1,41 @@
+      </div>
+
+      <div class="announcement-active">
+        <label for="active">Is Active:</label>
```
All free text needs to be translated: 
```html
<label for="active">{{_("Is Active:"}}</label>
```

pomegranited commented on openedx/core/djangoapps/announcements/templates/announcements/announcements_list.html:
```diff
@@ -0,0 +1,19 @@
+    </a>
+  {% empty %}
+    <div id="announcement-item">
+      No Announcements yet.
```
All free text needs to be translated:
```html
{{_("No Announcements yet."}}
```

pomegranited commented on openedx/core/djangoapps/announcements/templates/announcements/announcements_list.html:
```diff
@@ -0,0 +1,19 @@
+    </div>
+  {% endfor %}
+<a href="{% url 'announcements:announcements_create' %}">
+  <button>Add New Announcement</button>
```
All free text needs to be translated.

pomegranited commented on openedx/core/djangoapps/announcements/views.py:
```diff
@@ -0,0 +1,100 @@
+        return Fragment(html)
+
+
+class AnnouncementsEditView(EdxFragmentView):
```
I think adding/editing announcements should only be allowed to a limited number of users, say only users with `is_staff` or "course creator" status?  see e.g. `cms.djangoapps.contentstore.views/course._get_course_creator_status`, though this is a terrible place for this method to be if we want to check it from `openedx`.

kaizoku replied:
Ah, I mistakenly thought studio login was restricted to staff already.
I've used require_global_staff from util.views to fix this.

pomegranited commented on openedx/core/djangoapps/announcements/templates/announcements/announcements_edit.html:
```diff
@@ -0,0 +1,41 @@
+{% block content %}
+<div class="wrapper-content wrapper announcements-editor">
+  <section class="content activation is-active">
+    <form method="post" action="">
```
form `action` should be `announcements_list`.

pomegranited commented on openedx/core/djangoapps/announcements/templates/announcements/announcements_list.html:
```diff
@@ -0,0 +1,19 @@
+<div class="announcements-list">
+  {% for announcement in announcement_list %}
+    <a href="{% url 'announcements:announcements_edit' pk=announcement.pk %}">
+      <div class="announcement-item {% if announcement.active %}announcement-active{% endif %}">
```
The "active" class state with styling is too implicit, and not accessible. 
Better to provide a table that labels each announcement as "active" or "inactive".

Also need to allow for pagination here.

pomegranited commented on openedx/core/djangoapps/announcements/views.py:
```diff
@@ -0,0 +1,100 @@
+        """
+        Render page which lists all announcements
+        """
+        announcements = Announcements.objects.all()
```
Needs to be paginated.

pomegranited commented on common/djangoapps/student/views.py:
```diff
@@ -864,7 +865,11 @@ def dashboard(request):
     display_sidebar_on_dashboard = len(order_history_list) or verification_status in valid_verification_statuses
 
+    # display any announcements
+    announcements = json.dumps([announcement.content for announcement in Announcements.objects.filter(active=True)])
```
Also needs to be paginated, or limited by a configurable maximum number.

pomegranited commented on openedx/core/djangoapps/announcements/templates/announcements/announcements_edit.html:
```diff
@@ -0,0 +1,41 @@
+{% block title %}{{_("Edit Announcement")}}{% endblock %}
+
+{% block content %}
+<div class="wrapper-content wrapper announcements-editor">
```
Would be much nicer/cleaner if the adding/editing could be done inline on the list page, without having to load another page.

kaizoku commented:
- Yes, I've changed the TinyMCE commit to use the same version of TinyMCE, just with the extra plugins bundled.
- I wasn't sure if this should be on the front page, but it may help the UI be more integrated and accessible. I've started working on that, but it's a bit more complicated than the simple UI I have built out already, so it will take a bit longer.
- Ah, that seems like a more appropriate location. I've moved it to `openedx/features`
- Right, Thanks, I've added the migration step to my testing steps.
- Yes, I'll add tests as well.

pomegranited commented:
Hey @edx/edx-ux-team! @e've started implementing this Announcements feature, which @bdero started discussing with @marcotuts .  During that discussion, we determined that placing this feature in the GUI is going to be a bit awkward, since it's for site-wide announcements shown (initially) in just the Learner Dashboard.  We ended up deciding to put the editing panes in Studio, but now are unsure where in Studio it should go.

Does it make sense for it to be another tab on the Studio front page?  It's disabled by a feature flag, so won't affect sites that don't enable it.

pomegranited commented on openedx/features/announcements/views.py:
```diff
@@ -80,3 +82,31 @@ def get(self, request, **kwargs):
+        """
+        page = kwargs.get("page", 1)
+        announcements = Announcement.objects.filter(active=True)
+        paginator = Paginator(announcements, 5)
```
`5` should be configurable.

kaizoku replied:
Ah, good idea. I've added a setting for this under settings.ANNOUNCEMENT_SETTINGS.

pomegranited commented:
jenkins run all

pomegranited commented:
jenkins ok to test

pomegranited commented on openedx/features/announcements/static/announcements/jsx/announcements.jsx:
```diff
@@ -0,0 +1,91 @@
+import ReactDOM from 'react-dom';
+import PropTypes from 'prop-types';
+
+class Announcement extends React.Component {
```
This code needs to be tested too.  See https://github.com/edx/paragon/blob/master/src/StatusAlert/StatusAlert.test.jsx for example.

pomegranited replied:
JSX tests still need to be added?

kaizoku commented on openedx/features/announcements/views.py:
```diff
@@ -0,0 +1,114 @@
+        return Fragment(html)
+
+
+class AnnouncementsEditView(EdxFragmentView):
```
Hmm, I was using the Django generic views initially, but as you mentioned, they can't render to a fragment view.

Unfortunately I can't subclass both the generic views, and EdxWebFragment view.
I think this is due to [FragmentView overriding the metaclass](https://github.com/edx/web-fragments/blob/6ed3f4e451e4860dbed8ee1a04ffa026e500ff3b/web_fragments/views.py#L20) of the django.views.generic.View from which it inherits. I get this traceback when I attempt it:
```
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "/edx/app/edxapp/edx-platform/openedx/features/announcements/views.py", line 117, in <module>
    class AnnouncementFormView(EdxFragmentView, FormView):
  File "/edx/app/edxapp/venvs/edxapp/lib/python2.7/abc.py", line 87, in __new__
    cls = super(ABCMeta, mcls).__new__(mcls, name, bases, namespace)
TypeError: Error when calling the metaclass bases
    metaclass conflict: the metaclass of a derived class must be a (non-strict) subclass of the metaclasses of all its bases
```

pomegranited replied:
@kaizoku Bummer!

pomegranited replied:
@kaizoku I was able to remove a lot of duplicated logic by combining the `EdxFragmentView` and `django.views.generic.list.ListView` like this:

```python
class AnnouncementsView(EdxFragmentView, ListView):
    """
    View showing a list of all announcements.
    """
    template_name = "announcements/announcements_list.html"
    paginate_by = settings.get('ANNOUNCEMENT_SETTINGS', {}).get('ANNOUNCEMENTS_PER_PAGE', 3)
    model = Announcement
    context_object_name = 'announcement_list'
    object_list = Announcement.objects.all()
    
    def render_to_fragment(self, request, **kwargs):
        """
        Render page which lists all announcements
        """
        html = render_to_string(self.template_name, self.get_context_data(), request=request)
        return Fragment(html)
```

You'll need to update your [pagination variables in the template](https://github.com/edx/edx-platform/pull/16496/files#diff-21ac363084395cedba9908f97846ae3fR37), as shown [on this tutorial](https://simpleisbetterthancomplex.com/tutorial/2016/08/03/how-to-paginate-with-django.html) (see user_list.html).

Please apply the same logic to the other views.

kaizoku commented on openedx/features/announcements/tests/test_announcements.py:
```diff
@@ -0,0 +1,69 @@
+        """Ensures that the default settings effectively disables the feature"""
+        response = self.client.get('/dashboard')
+        self.assertNotIn('AnnouncementsView', response.content)
+        self.assertNotIn('<div id="announcements"', response.content)
```
Added these tests apart from the tests for Edit and Delete, still working on those.
I hadn't implemented them to begin with since I thought they might change if we move the editor around in the studio anyhow.

kaizoku replied:
Scratch that, you just helped me fix the tests for these. I pushed them as well.

pomegranited replied:
+1

pomegranited commented on lms/envs/common.py:
```diff
@@ -64,6 +64,10 @@
 
+ANNOUNCEMENT_SETTINGS = {
+    'ANNOUNCEMENTS_PER_PAGE': 5,
+}
```
@kaizoku You should also use a variable when paginating the Studio views, so you'll need to include this stanza in `cms/envs/common.py` too (yes, ironic that "common" doesn't apply across both systems).

edx-webhook commented:
Thanks for the pull request, @kaizoku! I've created [OSPR-2011](https://openedx.atlassian.net/browse/OSPR-2011) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).



If you like, you can add yourself to the [AUTHORS](https://github.com/edx/edx-platform/blob/master/AUTHORS)      file for this repo, though that isn't required.     Please see the [CONTRIBUTING](https://github.com/edx/edx-platform/blob/master/CONTRIBUTING.rst)     file for more information.

pomegranited commented:
@nedbat  Could we get approval to run jenkins tests for this PR?

pomegranited commented:
@kaizoku :+1: once jenkins tests pass.

- [x] I tested this on my devstack, with and without `FEATURES.ENABLE_ANNOUNCEMENTS`.
- [x] I read through the code
- [x] I checked for accessibility issues -- uses accessible React components from [edx/paragon](https://github.com/edx/paragon)
- [ ] Includes documentation -- will need docs added to edx/documentation CC @edx/doc

nedbat commented:
jenkins ok to test

katymyw commented:
@pomegranited -- In reviewing with @sstack22, we think this makes sense to appear on the Maintenance page in Studio. 

For edx.org use, we'll want to make sure that only global staff can update these announcements.

kaizoku commented:
@katymyw Thank you for the clarification. I'll move the editor into the Maintenance page and update the templates so they use the appropriate classes for the maintenance stylesheets.

srpearce commented:
@grantgoodmanedX This looks like something we may need to document for Open edX.

grantgoodmanedX commented:
@srpearce I created DOC-3856 for the documentation.

nedbat commented:
@katymyw Is product satisfied with the state of this pull request?

@kaizoku You will need to rebase (at least) to get the tests to pass.

katymyw commented:
@kaizoku -- Can you clarify whether the latest version addresses the comments above (on maintenance page, ability to limit to global staff)?

kaizoku commented:
@katymyw & @nedbat Sorry I haven't gotten back to this, it's been a busy couple of months.
I have time scheduled to respond to the review this week.
I'll let you know when it's ready to review again.

mduboseedx commented:
@kaizoku Have you had a chance to review this again?

kaizoku commented:
@mduboseedx I've just added a basic editor in the maintenance view from the studio.
I made sure to restrict the announcement editing pages to global staff.

I have yet to add tests for these views, and the style for the editor is a bit rudimentary.
I'll comment again here when I've added tests and updated the styles in the editor.

kaizoku commented:
@mduboseedx I've updated the stylesheets and fixed the tests for the announcement editor in the Maintenance Dashboard.

I had to alter some of the existing tests for the maintenance dashboard, since they weren't working with multiple maintenance pages. The newer version of ddt [has a function `ddt.idata` which takes an iterator](https://github.com/datadriventests/ddt/blob/88bd6225661ad581aac8d8d8bdb58603c9b1b8cb/ddt.py#L69) and is ideally what would be used here, but isn't available in 0.8.0 which is [the current version in master](https://github.com/edx/edx-platform/blob/db5042f08469c73d771b2b34cd0f2546a9f464c2/requirements/edx/testing.txt#L79).

pomegranited commented on cms/djangoapps/maintenance/views.py:
```diff
@@ -211,3 +226,80 @@ def post(self, request):
+    @method_decorator(require_global_staff)
+    def get(self, request, *args, **kwargs):
+        if "limit" in request.GET.keys():
+            self.paginate_by = request.GET.get("limit")
```
Need to enforce an upper limit here, else people could bring the database to its knees by creating and requesting millions of announcements on a page.

kaizoku replied:
I added `max(..., 50)`. Does that seem like a reasonable limit? Perhaps it should be a bit higher even..

pomegranited commented on cms/templates/maintenance/_announcement_edit.html:
```diff
@@ -0,0 +1,57 @@
+    <div class="container studio-view maintenance-form">
+        <form class="form-create" method="post" action="${action_url}">
+            <div class="wrapper-form announcement-container">
+                <input type="hidden" name="csrfmiddlewaretoken" value="${csrf_token}"/>
```
These form templates can be greatly simplified if you use the built-in Django form rendering, e.g.,
 
```diff
diff --git a/cms/templates/maintenance/_announcement_edit.html b/cms/templates/maintenance/_announcement_edit.html
index 8915dc4069..4ee4019db9 100644
--- a/cms/templates/maintenance/_announcement_edit.html
+++ b/cms/templates/maintenance/_announcement_edit.html
@@ -19,17 +19,7 @@ from openedx.core.djangolib.markup import HTML, Text
             <div class="wrapper-form announcement-container">
                 <input type="hidden" name="csrfmiddlewaretoken" value="${csrf_token}"/>
                 <div class="announcement-item">
-                    <textarea name="content" class="announcement-content">
-                        ${getattr(announcement, "content", "") | n}
-                    </textarea>
-                    <div class="actions-list">
-                        <label for="active">${_('Is Active')}</label>
-                        <input type="checkbox" name="active" class="announcement-active"
-                            %if announcement.active:
-                                checked
-                            %endif
-                         />
-                    </div>
+                ${form.as_ul() | n}
                 </div>
                 <div class="actions">
                     <button type="submit" class="action action-primary">${_('Save')}</button>
```

pomegranited replied:
+1

pomegranited commented on cms/djangoapps/maintenance/views.py:
```diff
@@ -211,3 +226,80 @@ def post(self, request):
+    def get(self, request, *args, **kwargs):
+        super(AnnouncementCreateView, self).get(request, *args, **kwargs)
+        context = self.get_context_data()
+        context['action_url'] = reverse('maintenance:announcement_create')
```
But if all you're doing is adding something to the context dict, you can do that by overriding `get_context_data()`, and let `super.get()` be used as-is.

pomegranited commented on cms/djangoapps/maintenance/views.py:
```diff
@@ -211,3 +226,80 @@ def post(self, request):
+    model = Announcement
+    fields = ['content']
+    success_url = '/maintenance/announcements'
+    template = '/maintenance/_announcement_edit.html'
```
You can greatly simplify these edit views and their templates by using `template_name` here instead.  See e.g. [ApiRequestView](https://github.com/edx/edx-platform/blob/8c66d867044295eac4ded838028e43e8521bfc9f/openedx/core/djangoapps/api_admin/views.py#L28), and its template, [api_access_request_form.html](https://github.com/edx/edx-platform/blob/6d2468d0dcf782a7561ecc61946146d6fae2762f/lms/templates/api_admin/api_access_request_form.html#L19).

To editing with a textarea, use `widget=forms.Textarea` in the custom form field  (cf https://stackoverflow.com/a/430642/4302112).

pomegranited commented:
@kaizoku Thanks for those updates!  The code is much leaner now.  But I'm unable to create a new Announcement.. the Save button throws a JS error and then doesn't post: `An invalid form control with name='content' is not focusable.`

Also cannot save an announcement without clicking the "Enable" box -- this goes for creating new and editing existing announcements.  GUI shows this:
![screen shot 2018-05-04 at 4 17 25 pm](https://user-images.githubusercontent.com/7556571/39615674-a7d91bc8-4fb6-11e8-96e6-1de3d9c054ff.png)

kaizoku commented:
@pomegranited Ah, thanks for the find there. The fields for the django form were still marked required (the default), and that was messing up the javascript.
Oddly I didn't notice since it was actually working OK in my version of firefox, but I was seeing the same issue you did in chromium.

I just pushed another commit to make those form fields not required,and now it's working for me in both firefox and chromium.

pomegranited commented on cms/djangoapps/maintenance/views.py:
```diff
@@ -211,3 +227,77 @@ def post(self, request):
+    """
+    model = Announcement
+    form_class = AnnouncementForm
+    success_url = '/maintenance/announcements'
```
Should use `reverse_lazy('maintenance:announcement_index')` instead of hardcoded path (cf [django docs for why `reverse_lazy`](https://docs.djangoproject.com/en/1.11/topics/class-based-views/generic-editing/#model-forms)).

pomegranited commented on cms/djangoapps/maintenance/views.py:
```diff
@@ -211,3 +227,77 @@ def post(self, request):
+    """
+    model = Announcement
+    form_class = AnnouncementForm
+    success_url = '/maintenance/announcements'
```
[ditto](https://github.com/edx/edx-platform/pull/16496/files#diff-af4e7f6c351fdaebd9a7faadbc90ca12R273)

pomegranited commented on cms/djangoapps/maintenance/views.py:
```diff
@@ -211,3 +227,77 @@ def post(self, request):
+    View for deleting an announcement.
+    """
+    model = Announcement
+    success_url = '/maintenance/announcements'
```
[ditto](https://github.com/edx/edx-platform/pull/16496/files#diff-af4e7f6c351fdaebd9a7faadbc90ca12R273)

pomegranited commented on cms/templates/maintenance/_announcement_delete.html:
```diff
@@ -0,0 +1,39 @@
+                </div>
+            </div>
+            <div class="actions">
+                <button type="submit" class="action action-primary">${_('Confirm')}</button>
```
Need to add `value="Confirm"` to the button, otherwise Django doesn't redirect back to the `success_url`.

nedbat commented:
This PR is mentioned on the Hawthorn wiki page (https://openedx.atlassian.net/wiki/spaces/COMM/pages/158045104/Hawthorn), but I'm not hopeful that it will be merged in time.

kaizoku commented:
@nedbat, The code is all ready for review. I've just rebased against master to ensure all the tests are still working.

Is there anything else I can do to help move this along?

nedbat commented:
jenkins run bokchoy

nedbat commented:
Josh, I wish I had better news for you.  As a thought experiment, and to help us get better as a platform, what extension API would we have to provide so that this feature could be built as an optional add-on?

antoviaque commented:
@nedbat Do you mean that the product review was negative? That would be quite disappointing to say the least, given that this product review started 8 month ago, and that more than 120h of work have been put into getting it upstream (not including the initial development time that we inherited).

nedbat commented:
@antoviaque I mean that the edX teams don't seem to be prioritizing this to get it done in time for Hawthorn.  I don't mean that it will never merge.  I am also disappointed by this.

antoviaque commented:
@nedbat Ah, I see - sorry about my answer then, I replied too fast. And thank you for the help on trying to get this merged on time for Hawthorn, even though it looks like it won't happen here - I appreciate it!

kaizoku commented:
@nedbat, I think this one would have been pretty difficult to create as an extension.
Since this requires adding HTML and a React component to the dashboard.

If there were some way to extend the dashboard without overriding the entire template through a theme, and incorporate webpack assets such as React (JSX) components, this may be be possible using an external djangoapp.

nedbat commented:
@arizzitano @nasthagiri this could be an interesting test case for an extensibility point.

pomegranited commented:
Hi @nedbat CC @mduboseedx  We've got another client who would like to use this feature.  Could we get some time from your team to help direct the architecture questions, so we can prepare this for review?

mduboseedx commented:
@edx/arch-review @nasthagiri Could you take a look at this from architecture perspective?

ormsbee commented:
@mduboseedx: @nasthagiri and I have looked this over. Architecturally, we're fine with the overall approach. The one request is that this PR uses Django pluggable apps style for configuration, like the grades app in edx-platform: https://github.com/edx/edx-platform/blob/master/lms/djangoapps/grades/apps.py

I realize we don't have a good solution for Webpack in Django plugins, so the existing setup there is fine.

@schenedx: Do you mind if I do the code review and merge on this feature?

schenedx commented:
@ormsbee Thank you for ask. Please go ahead.

ormsbee commented:
@mduboseedx, @pomegranited: I'll do review on this PR as soon as the Django plugin changes are made. If that's going to be difficult for any reason, please let me know.

Thank you.

kaizoku commented:
I've changed things around so this uses the django pluggable application configuration.
It still requires a modification to the dashboard to include the display UI, and also some modifications to the maintenance dashboard in studio, since it was requested to put the editor in that interface.

@ormsbee Does this sound alright to you? Please let me know if you'd like any other changes to this PR. Thanks in advance for reviewing it.

pomegranited commented:
Ach, looks like I neglected to [ping you specifically here](https://github.com/edx/edx-platform/pull/16496#issuecomment-415360015) @kaizoku -- would your client be willing to have the Announcements display below the Search box instead of above?

kaizoku commented:
Ah, sorry @pomegranited , I still should have caught that. They're not even using the dashboard search, so I doubt they will mind or notice.

I've just moved the announcements to show below the search.

pomegranited commented:
Thank you so much @kaizoku !  That helps me out heaps.

ormsbee commented:
Starting my review (though I might not be able to finish it until tomorrow).

ormsbee commented:
Sorry for the delay. I am batching up some comments in my review, but one out-of-band question: what's that standalone "p" doing near the bottom of the "Edit Announcement" page?

ormsbee commented:
I also get repeated 404 errors for assets on the `/maintenance/announcements/` page in devstack.

![announcements_404](https://user-images.githubusercontent.com/41625/46091545-114cfc00-c181-11e8-89c3-c4141a2a8965.png)

ormsbee commented:
Hmmm... maybe that 404 is just my devstack being borked somehow...

ormsbee commented:
Ah, the "p" is an indication of applied tags for the current cursor position. Sorry, still playing with this on my local machine.

mduboseedx commented:
@ormsbee Thanks for letting us know and for looking over this PR

kaizoku commented on cms/djangoapps/maintenance/views.py:
```diff
@@ -211,3 +227,77 @@ def post(self, request):
+    @method_decorator(require_global_staff)
+    def get(self, request, *args, **kwargs):
+        if "limit" in request.GET.keys():
+            self.paginate_by = max(request.GET.get("limit"), 50)
```
Ah, good catch. I'll remove that as it's not really necessary.

kaizoku commented on lms/templates/dashboard.html:
```diff
@@ -256,6 +256,10 @@ <h3>${course_dir}</h3>
 
+        % if settings.FEATURES.get('ENABLE_ANNOUNCEMENTS'):
+          <%include file='dashboard/_dashboard_announcements.html' args='announcements=announcements' />
+        % endif
```
Hmm, that might make more sense for accessibility. We had just put it there since it seems to fit well in the sidebar.

ormsbee replied:
@wittjeff: This was the PR I mentioned in the Accessibility chat room, where the announcements go after the course listing in the DOM ordering (this is the student dashboard). Is it sufficient to put role="alert" or role="status" on it, or is it critical that it also come before the course listing?

pomegranited replied:
a11y still needs to be addressed?

kaizoku commented on openedx/features/announcements/views.py:
```diff
@@ -0,0 +1,33 @@
+    """
+    model = Announcement
+    object_list = Announcement.objects.filter(active=True)
+    paginate_by = settings.FEATURES.get('ANNOUNCEMENTS_PER_PAGE', 3)
```
Right, that's odd. I've set this to 5 to match the default setting.

kaizoku commented on cms/templates/maintenance/_announcement_delete.html:
```diff
@@ -0,0 +1,39 @@
+          <div class="message has-warnings" aria-hidden="true">
+                <p class="warning">
+                    <span class="icon fa fa-warning"></span>
+                    Are you sure you want to delete this Announcement?
```
Looks like I missed a few of those, thanks for catching them.

kaizoku commented on cms/djangoapps/maintenance/tests.py:
```diff
@@ -260,3 +259,60 @@ def test_force_publish_dry_run(self):
+        url = reverse("maintenance:announcement_delete", kwargs={"pk": announcement.pk})
+        self.client.post(url)
+        result = Announcement.objects.filter(content="Test Edit Announcement").exists()
+        self.assertFalse(result)
```
I added some tests to try to load all of the pages as a regular user and ensure they return 403.

kaizoku commented:
Thanks for reviewing this @ormsbee. I updated the code in response to your comments, and left responses for those that may not require code changes.

kaizoku commented:
@ormsbee, @mduboseedx sorry to bug ya, but any idea when we can get another review on this one?

We have a few clients hoping to use this in the upgrade to hawthorn we're currently undergoing.

ormsbee commented:
@kaizoku: Once @wittjeff weighs in on the a11y issue, I'm fine with the rest of the PR.

pomegranited commented:
@wittjeff Do you know when you might have time to [comment on this a11y issue](https://github.com/edx/edx-platform/pull/16496#discussion_r224660186) ?  CC @mduboseedx @kaizoku

natabene commented:
@pomegranited  edX (@ormsbee ) will provide update shortly, it will include accessibility comments.

kaizoku commented:
@natabene, any word on the accesibility review for this one?

ormsbee commented:
Sorry, I was trying to see if there was a way to combine this with something else, but that's not going to happen. Relaying the a11y feedback:

1. Structurally, the clustering of messages is generally a good thing / best-practice for accessibility. The key objective then is discoverability for people who perhaps can't see it visually. The typical way to ensure discoverability is to put the block of messages at the top of the page.  Your approach is OK but I would suggest a couple of enhancements:
    a) Use Flexbox to position the block of messages to the right on wide viewports, or on top for narrow viewports.
    b) Add a link to the top of the page, before the Skip to main content link, that says "Skip to list of X messages". Use the same CSS (hide when not focused, show when focused) as the SkipNave link. Don't include the link if the list of message is empty.
2. **Optional**: Where you have `<div class="side-container">` please add `role="complementary"` and `aria-label="messages"`.
3. **Optional**: I suggest putting the messages in a UL/LI structure (without any additional styling). The extra structure helps screen reader users just a bit.
4. For your Next button, I suggest adding `<span class="sr-only">(X-Y of Z)</span>`
So that might read like "Next (6-8 of 8)" or "Next (6-10 of 23)".

kaizoku commented:
I'm trying to update this pull request. When I pushed commits to the associated branch, they naturally show up in that branch. Also, I can see the icon at the bottom here turn yellow and show "Checking for ability to merge automatically.." as though it's going to run tests again, but then it reverts to gray and the new commits don't show up in the commit list.

@ormsbee, could there be some webhooks failing that cause this to fail to update?

kaizoku commented:
Ah, github support just got back to me saying they've escalated to the engineering team, so this isn't such a common problem I guess. Sorry for the bother, I'll wait and see what they say.

ormsbee commented:
FYI @edx/testeng: Please see @kaizoku's note above ^

pomegranited commented on cms/templates/maintenance/_announcement_edit.html:
```diff
@@ -0,0 +1,47 @@
+            <div class="wrapper-form announcement-container">
+                <input type="hidden" name="csrfmiddlewaretoken" value="${csrf_token}"/>
+                <div class="announcement-item">
+                  ${form.as_p() | n} ## xss-lint: disable=mako-invalid-html-filter
```
nit: this `xss-lint` comment is rendering to the HTML:

![xss-comment](https://user-images.githubusercontent.com/7556571/51825196-2a82d680-2334-11e9-9371-88af60ed25e2.png)

kaizoku replied:
Ah, those have to be on their own line evidently. Thanks for catching that.

pomegranited commented on cms/templates/maintenance/_announcement_index.html:
```diff
@@ -0,0 +1,49 @@
+            % for announcement in announcement_list:
+                <div class="announcement-item">
+                    <div class="announcement-content">
+                        ${announcement.content | n} ## xss-lint: disable=mako-invalid-html-filter
```
[Ditto](https://github.com/edx/edx-platform/pull/16496/files#r251327119) for these `xss-lint` comments too.

pomegranited commented on cms/templates/maintenance/_announcement_delete.html:
```diff
@@ -0,0 +1,39 @@
+            <input type="hidden" name="csrfmiddlewaretoken" value="${csrf_token}"/>
+            <div class="announcement-item">
+                <div class="announcement-content">
+                    ${object.content | n} ## xss-lint: disable=mako-invalid-html-filter
```
[Ditto](https://github.com/edx/edx-platform/pull/16496/files#r251327119) for this `xss-lint` comment too.

pomegranited commented on cms/templates/maintenance/_announcement_index.html:
```diff
@@ -0,0 +1,49 @@
+            % for announcement in announcement_list:
+                <div class="announcement-item">
+                    <div class="announcement-content">
+                        ${announcement.content | n} ## xss-lint: disable=mako-invalid-html-filter
```
Would like a visible marker to show which announcements are disabled, and which are enabled, since this will get harder to manage as new announcements are added.

Maybe also if you sort them to show the active announcements first, in the order they'll be displayed on the LMS?

kaizoku replied:
I added a text that says "Active" or "Inactive", with a checkmark or an empty box. Does that seem like a good indicator? I wasn't sure exactly where or how to indicate this.

I've added a sort by active status on the queryset, so the inactive announcements will show up last.

pomegranited replied:
Beaut! 👍

pomegranited commented on cms/templates/maintenance/_announcement_edit.html:
```diff
@@ -0,0 +1,47 @@
+              "advlist autolink lists link image charmap print preview anchor",
+              "searchreplace visualblocks code fullscreen",
+              "insertdatetime media table contextmenu paste"
+          ],
```
Since `lists` don't render very nicely or consistently between the CMS, LMS, and the `preview` and `fullscreen` views are also misleading, maybe it's better to just remove all of these options?

This would also reduce the number of tinymce plugins added.

* CMS:
  ![cms lists](https://user-images.githubusercontent.com/7556571/51825785-72eec400-2335-11e9-9733-136c9ac5cad2.png)
* LMS:
  ![lms lists](https://user-images.githubusercontent.com/7556571/51825793-76824b00-2335-11e9-9c20-e704abc2ccf2.png)
* Preview: 
  ![preview](https://user-images.githubusercontent.com/7556571/51826016-f5778380-2335-11e9-917b-4c0f1b52ad25.png)
* Fullscreen: 
  ![fullscreen](https://user-images.githubusercontent.com/7556571/51826024-fad4ce00-2335-11e9-8c04-e61af304ff0c.png)
  Also, I couldn't find a way to get back out of Fullscreen?

kaizoku replied:
Quite right, I've removed those plugins as they aren't really essential.
Those were only included because I copied the iist of plugins that were in use before.
I've pruned them now, and regenerated the TinyMCE plugins so they're not included.

pomegranited commented on cms/templates/maintenance/_announcement_index.html:
```diff
@@ -0,0 +1,49 @@
+            <a href="${ reverse('maintenance:announcement_create') }">
+                <button class="action action-primary">${_('Create New')}</button>
+            </a>
+            % if is_paginated:
```
Pagination doesn't seem to be working in the CMS view?  I've created 11 announcements, and they're all showing up.

kaizoku replied:
Ah whoops, fixed that one. I guess I unintentionally checked in a `self.paginate_by = 50` in the `get()` method for the view from when I was testing things.

pomegranited replied:
Working now, thanks! +1

pomegranited commented on scripts/xsslint_thresholds.json:
```diff
@@ -9,7 +9,7 @@
         "javascript-jquery-prepend": 7,
         "mako-html-entities": 0,
-        "mako-invalid-html-filter": 9,
+        "mako-invalid-html-filter": 13,
```
Why did this limit have to be increased?

kaizoku replied:
This was for the editor in the maintenance view, since we're showing the announcement unescaped for TinyMCE to edit. It looks like these have moved to the inline overrides though, so I've removed this limit increase.

pomegranited replied:
+1

kaizoku commented:
@pomegranited, for some reason I can't reply on the last few comments, so I'll reply here.
I'm not sure why it was still only showing the include for `dashboard/_dashboard_announcements.html`, I had included a skip link when you reviewed this, and I'm seeing it in the "Files changed" tab. Was that the a11y you meant I should address?

Ah right, there are JSX tests now in the platform. ;) I've added some tests for both the announcement list and the skip link React components.

This is ready for review apart from the failing a11y test for the skip link, though I can't figure out why it's failing. It doesn't fail in my devstack. I can reproduce the failure by removing the `tabindex="-1"` from the [skip link's target](https://github.com/edx/edx-platform/blob/e02366914c2d11bbee3230f7b61400bd8a0914bb/lms/templates/dashboard/_dashboard_announcements.html#L3), but I'm not sure why it's failing on jenkins, but not on my local machine. Any ideas @pomegranited ?

pomegranited commented:
jenkins run bokchoy

kaizoku commented:
@ormsbee I've addressed the a11y issues, though there's a test failure since the test suite seems to expect that the first skip link is to the main content. The a11y review said that the announcements skip link should come before the main content link though.
Is it alright if we move the announcements skip link to after the main skip link?

Apart from that one point, this is ready for review. I believe the bokchoy tests that's failing is a flaky test. It's in an unrelated feature and sometimes was failing, and sometimes succeeding before.

wittjeff commented:
I'd prefer to fix the test on this one (to allow the skip announcements
link first, for discoverability).  I think the regular SkipNav link is a
cue the says "all is normal in the header, go ahead and jump ahead) but we
really don't want them to think that.

On Mon, Feb 11, 2019 at 2:29 AM Josh <notifications@github.com> wrote:

> @ormsbee <https://github.com/ormsbee> I've addressed the a11y issues,
> though there's a test failure since the test suite seems to expect that the
> first skip link is to the main content. The a11y review said that the
> announcements skip link should come before the main content link though.
> Is it alright if we move the announcements skip link to after the main
> skip link?
>
> Apart from that one point, this is ready for review. I believe the bokchoy
> tests that's failing is a flaky test. It's in an unrelated feature and
> sometimes was failing, and sometimes succeeding before.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/edx/edx-platform/pull/16496#issuecomment-462237873>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABwz81zVxZ40QXF5uyTornQ6pWsVIjinks5vMRvAgaJpZM4QWvET>
> .
>

ormsbee commented:
@kaizoku: Please fix the test as @wittjeff suggests, squash, make sure to have a descriptive commit message with the background -- and then we'll get this thing merged. Thank you.

pomegranited commented on lms/templates/dashboard.html:
```diff
@@ -272,9 +272,11 @@ <h3>${course_dir}</h3>
 
         % if settings.FEATURES.get('ENABLE_ANNOUNCEMENTS'):
```
@kaizoku Wow, weird that you had to do this, but I guess it's something about the template `<%block>` tags?

To make this look less confusing, could you move this line down to avoid wrapping this in two `% if settings.FEATURES.get('ENABLE_ANNOUNCEMENTS')` blocks? i.e.

```
          <%block name="skip_links">
            % if settings.FEATURES.get('ENABLE_ANNOUNCEMENTS'):
              <a class="nav-skip sr-only sr-only-focusable" href="#announcements" id="announcements-skip">${_("Skip to list of announcements")}</a>
            % endif
          </%block>
        % if settings.FEATURES.get('ENABLE_ANNOUNCEMENTS'):
          <%include file='dashboard/_dashboard_announcements.html' />
        % endif
```

kaizoku replied:
Yeah, I guess the blocks are rendered prior to the conditionals being evaluated or something?
This is especially strange because it's contrary to the Mako docs:
> Blocks also work in iterations, conditionals, just like defs:
> ```
> % if some_condition:
>     <%block>condition is met</%block>
> % endif
> ```

Though I guess this is not true for named blocks. I just tested, and while regular blocks work as expected in conditionals, named blocks don't.

Your refactor is definitely clearer though, I've changed it to match that pattern. Thanks!

kaizoku commented:
Thanks @ormsbee . I've fixed those tests, squashed, and updated the commit message.
Is there anything missing from the description there?

kaizoku commented:
jenkins run python

edx-webhook commented:
@kaizoku 🎉 Your pull request was merged!

Please take a moment to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/edx-platform&entry.867055334=https://github.com/edx/edx-platform/pull/16496&entry.1484655318=https://github.com/kaizoku&entry.752974735=2017-11-08+17:19&entry.1917517419=2019-03-18+20:04&entry.2133058324=Yes) so we can improve your experience in the future.
